server/src/main/java/org/apache/cassandra/sidecar/tasks/PeriodicTaskExecutor.java

```diff
@@ -130,22 +147,33 @@ private void schedule(PeriodicTaskKey key, long priorExecDurationMillis, long de
      */
     private void executeAndScheduleNext(PeriodicTaskKey key, long execCount)
     {
+        Promise<Void> runPromise = Promise.promise();
+        if (activeRuns.computeIfAbsent(key, k -> runPromise.future()) != runPromise)
+        {
+            LOGGER.debug("already active. task='{}' execCount={}", key, execCount);
+            return;
+        }
         long startTime = System.nanoTime();
-        internalPool.<Void>executeBlocking(promise -> executeInternal(promise, key, execCount), false)
-                    .onComplete(ignored -> {
-                        LOGGER.debug("Task run finishes. task='{}' execCount={}", key, execCount);
-                        // schedule the next run iff the task is not killed
-                        if (poisonPilledTasks.remove(key))
+        internalPool.<Result>executeBlocking(promise -> executeInternal(promise, key, execCount), false)
+                    .onComplete(outcome -> {
+                        LOGGER.debug("Task run finishes. task='{}' outcome={} execCount={}", key, outcome, execCount);
+                        runPromise.complete(); // mark the completion, regardless of the result from last run
+                        activeRuns.remove(key);
+
+                        DurationSpec delay;
+                        long priorExecutionDurationMillis;
+                        if (outcome.result() == Result.RESCHEDULED)
+                        {
+                            priorExecutionDurationMillis = 0;
+                            delay = key.task.initialDelay();
```

Comment:
Why do we always use the `initialDelay` here? What if the task has run 20 times already - do we still want to use `initialDelay`? Can we use `execCount` here to decide which delay to use?

Reply:
Yes. It is to implement the semantics of "reschedule". When it is rescheduled, the first run the rescheduled task take the `initialDelay`. It is consistent with the prior `reschedule` method.